qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/SingleBrokerCorruptionTest.java

```diff
@@ -0,0 +1,144 @@
+import io.zeebe.model.bpmn.Bpmn;
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Optional;
+import org.agrona.MutableDirectBuffer;
+import org.agrona.concurrent.UnsafeBuffer;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class SingleBrokerCorruptionTest {
+
+  private static final int PARTITION = 1;
+  @Rule public final ClusteringRule clusteringRule = new ClusteringRule(1, 1, 1);
+
+  @Test
+  public void shouldDetectCorruptionOnStart() throws IOException {
+    // given
+    final int nodeId = clusteringRule.getLeaderForPartition(PARTITION).getNodeId();
+    clusteringRule
+        .getClient()
+        .newDeployCommand()
+        .addProcessModel(
+            Bpmn.createExecutableProcess("process").startEvent().done(), "process.bpmn")
+        .send()
+        .join();
+
+    final Path directory = clusteringRule.getSegmentsDirectory(clusteringRule.getBroker(nodeId));
+    final Optional<File> optLog =
+        Arrays.stream(directory.toFile().listFiles())
+            .filter(f -> f.getName().equals("raft-partition-partition-1-1.log"))
+            .findFirst();
+
+    // when
+    assertThat(optLog).isPresent();
+    final File log = optLog.get();
+
+    clusteringRule.stopBroker(nodeId);
+    assertThat(LogCorrupter.corruptRecord(log, 1)).isTrue();
+
+    // then
+    assertThatThrownBy(() -> clusteringRule.startBroker(nodeId))
+        .hasRootCauseInstanceOf(CorruptedLogException.class);
+  }
+
+  @Test
+  public void shouldHandlePartialWrite() throws IOException {
+    // given
+    final int nodeId = clusteringRule.getLeaderForPartition(PARTITION).getNodeId();
+    clusteringRule
+        .getClient()
+        .newDeployCommand()
+        .addProcessModel(
+            Bpmn.createExecutableProcess("process").startEvent().done(), "process.bpmn")
+        .send()
+        .join();
+
+    final Path directory = clusteringRule.getSegmentsDirectory(clusteringRule.getBroker(nodeId));
+    final Optional<File> optLog =
+        Arrays.stream(directory.toFile().listFiles())
+            .filter(f -> f.getName().equals("raft-partition-partition-1-1.log"))
+            .findFirst();
+
+    // when
+    assertThat(optLog).isPresent();
+    final File log = optLog.get();
+
+    clusteringRule.stopBroker(nodeId);
+    final long lastIndex = getLastIndexWrittenToMetastore(nodeId);
+    assertThat(LogCorrupter.corruptRecord(log, lastIndex)).isTrue();
+    changeMetastoreIndex(nodeId, lastIndex - 1);
+
+    // then
+    clusteringRule.startBroker(nodeId);
+    clusteringRule
+        .getClient()
+        .newCreateInstanceCommand()
+        .bpmnProcessId("process")
+        .latestVersion()
+        .send()
+        .join();
+  }
+
+  private long getLastIndexWrittenToMetastore(final int nodeId) throws IOException {
```

Comment:
Can we move this test to Raft? Then we don't have to expose the implementation of metastore. We can use `metaStore.loadLastIndex` and `metastore.storeLastIndex` directly.